### DISCUSSION TITLE
haveIBeenPwned.esm.js error in angular

### DISCUSSION THREAD

**Discussion by TYee2011** [zxcvbn-ts/zxcvbn#152] [Q&A] [answered]
I am creating a password meter component in angular 11, part of the requirements is to add the matcher-pwned library, which I did, but when I try to do ng serve, I am getting compiler errors
```
Error: ./node_modules/@zxcvbn-ts/matcher-pwned/dist/haveIBeenPwned.esm.js
Module not found: Error: Can't resolve 'crypto' in 'C:\Work\sprapp\bostonsoftware.sprapp\node_modules\@zxcvbn-ts\matcher-pwned\dist'
```
any suggestions on how to resolve this? I tried searching on stack overflow but could not find a valid answer.
My application/ component worked fine before trying to add matcher-pwned package via npm

**Comment by MrWook** [comment, marked as answer]
Interesting 🤔 
It should distinguish between node and a browser build and it should do it on only if you use zxcvbn-ts.
Are you triggering the matcher on serve?

**Reply by TYee2011** [reply]
no, not triggering it on ng serve, it is part of my angular component, I set the options and matcher up on the nginit, which should trigger after the components view gets rendered, it seems to be just an issue with the matcher because when I comment out the matcherPwnedFactory, it works fine

`import { Component, Input, OnInit, OnChanges, SimpleChanges } from '@angular/core';
import { zxcvbn, zxcvbnOptions } from '@zxcvbn-ts/core';
import zxcvbnCommonPackage from '@zxcvbn-ts/language-common';
import zxcvbnEnPackage from '@zxcvbn-ts/language-en';
import matcherPwnedFactory from '@zxcvbn-ts/matcher-pwned';

@Component({
  selector: 'app-password-meter',
  templateUrl: './password-meter.component.html',
  styleUrls: ['./password-meter.component.scss']
})
export class PasswordMeterComponent implements OnInit, OnChanges {
result;
suggestions;
@Input() password;
  constructor() { }

  ngOnInit(): void {
    const options = {
      translations: zxcvbnEnPackage.translations,
      graphs: zxcvbnCommonPackage.adjacencyGraphs,
      dictionary: {
        ...zxcvbnCommonPackage.dictionary,
        ...zxcvbnEnPackage.dictionary,
      },
    };
    const matcherPwned = matcherPwnedFactory(fetch, zxcvbnOptions);
    zxcvbnOptions.addMatcher('pwned', matcherPwned);

    zxcvbnOptions.setOptions(options);
    this.result = zxcvbn(this.password);
    this.suggestions = this.result.feedback.suggestions;
  }

  ngOnChanges(changes: SimpleChanges): void {
     this.result = zxcvbn(this.password);
    this.suggestions = this.result.feedback.suggestions;
  }

  getClass() {
    return `strength-${this.result.score}`;
  }



}
`

**Reply by MrWook** [reply]
Hm... i will investigate this.

You are using `this.result = zxcvbn(this.password);
this.suggestions = this.result.feedback.suggestions;` in the `ngOnInit ` block.
If you remove it, does the error still exist?

**Reply by TYee2011** [reply]
yes, I still get the same error if I comment out those 2 lines. It only compiles when I comment out the matcherPwnedFactory

**Reply by MrWook** [reply]
Hm i could reproduce this issue with `ng new some-project-name`  and adding `zxcvbn-ts`.

I have no idea wtf angular is doing...
1. It should not execute this code piece
2. it is type checking the node_modules/zxcvbn-ts folder for some reason
3. It tries to compile the zxcvbn-ts src files directly from the node_modules_folder

I will test a few more thing, but i guess i need to make a separated packages for node and the brower 🤔

**Reply by TYee2011** [reply]
Any update on this? I am hoping to get this integrated into my code soon, since the rest of my component is done, just trying to add haveIbeenpwnd.
Thanks

**Reply by MrWook** [reply]
I found some relating issues like this one https://github.com/facebook/create-react-app/issues/11756

It seems like webpack 5, which is used by angular 11, is handling node packages differently than before.
Previously webpack just throw some polifills in the frontend bundle even if it is inside an `if` statement.

It should work if you add this to your package.json
```
  "browser": {
    "crypto": false
  }
```

**Reply by TYee2011** [reply]
Thank you, I did try that, but for some reason, the app still had the same error. But in any case, yesterday I decided to look over haveibeenpwned's documentation and wrote a function for my meter component that will trigger the password range endpoint after focusing off the input field. I might revisit your functionality once we are ready to update to the latest version of angular

**Reply by MrWook** [reply]
Hm it should had worked. In my test project with angular it fixed the issue 🤔
